File: abb_program.py
from tad import *


choice = 1
def preorderWithBracket(abb):
	print('(', end="")
	if not isEmptyTree(abb):
		if not isEmptyTree(leftSon(abb)) or not isEmptyTree(rightSon(abb)):
			print(' ',key(abb),' ', end="")
			preorderWithBracket(leftSon(abb))
			print(' ', end="")
			preorderWithBracket(rightSon(abb))
		else:
			print(' ',key(abb),' ', end="")
	print(')', end="")
		
if choice==1:
	tree = emptyTree(None)
	tree = insertKey(tree,4)
	tree = insertKey(tree,4)
	tree = insertKey(tree,2)
	tree = insertKey(tree,6)
	tree = insertKey(tree,1)
	tree = insertKey(tree,3)
	tree = insertKey(tree,5)
	tree = insertKey(tree,7)

	preorderWithBracket(tree)
	print()
	
	print('search 1...',key(searchKey(tree,1)))
	print('search 2...',key(searchKey(tree,2)))
	print('search 3...',key(searchKey(tree,3)))
	print('search 4...',key(searchKey(tree,4)))
	print('search 5...',key(searchKey(tree,5)))
	print('search 6...',key(searchKey(tree,6)))
	print('search 7...',key(searchKey(tree,7)))

	print('remove 5...')
	tree = eraseKey(tree,5)
	preorderWithBracket(tree)
	print()

	print('remove 6...')
	eraseKey(tree,6)
	preorderWithBracket(tree)
	print()

	print('remove 4...')
	eraseKey(tree,4)
	preorderWithBracket(tree)
	print()

	print('remove 2...')
	eraseKey(tree,2)
	preorderWithBracket(tree)
	print()

File: oCaml/abb_program.ml
let rec preorderWithBracket abb = 
    print_string "(";
    if not (isEmptyTree abb ) then
        if not (isEmptyTree !(leftSon abb)) || not (isEmptyTree !(rightSon abb)) then
            begin
                print_string " ";
                print_int (key abb);
                print_string " ";
                preorderWithBracket !(leftSon abb);
                print_string " ";
                preorderWithBracket !(rightSon abb);

            end
        else begin
            print_string " ";
            print_int (key abb);
            print_string " ";

        end;
        print_string ")";;

let test i = match i with
    0->()
    |1->
        let abb = ref emptyTree in
        insertKey abb 4;
        insertKey abb 4;
        insertKey abb 2;
        insertKey abb 6;
        insertKey abb 1;
        insertKey abb 3;
        insertKey abb 5;
        insertKey abb 7;


        preorderWithBracket !abb;
        print_string "\n";

        print_string " Buscar 1... ";
        print_int (key(searchKey abb 1));
        print_string "\n";

        print_string " Buscar 2... ";
        print_int (key(searchKey abb 2));
        print_string "\n";

        print_string " Buscar 3... ";
        print_int (key(searchKey abb 3));
        print_string "\n";

        print_string " Buscar 4... ";
        print_int (key(searchKey abb 4));
        print_string "\n";

        print_string " Buscar 5... ";
        print_int (key(searchKey abb 5));
        print_string "\n";

        print_string " Buscar 6... ";
        print_int (key(searchKey abb 6));
        print_string "\n";

        print_string " Buscar 7... ";
        print_int (key(searchKey abb 7));
        print_string "\n";

        print_string "remove 5...";
        eraseKey abb 5;
        preorderWithBracket !abb;
        print_string "\n";

        print_string "remove 6...";
        eraseKey abb 6;
        preorderWithBracket !abb;
        print_string "\n";

        print_string "remove 4...";
        eraseKey abb 4;
        preorderWithBracket !abb;
        print_string "\n";

        print_string "remove 2...";
        eraseKey abb 2;
        preorderWithBracket !abb;
        print_string "\n";
        ();;
